File: Example/Tests/NYTPhotosViewControllerTests.m
//
//  NYTPhotosViewControllerTests.m
//  NYTPhotoViewer
//
//  Created by Brian Capps on 2/24/15.
//  Copyright (c) 2015 NYTimes. All rights reserved.
//

@import UIKit;
@import XCTest;

#import <NYTPhotoViewer/NYTPhotosViewController.h>
#import "NYTExamplePhoto.h"

@interface NYTPhotosViewControllerTests : XCTestCase

@end

@implementation NYTPhotosViewControllerTests

- (void)testPanGestureRecognizerExistsAfterInitialization {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    XCTAssertNotNil(photosViewController.panGestureRecognizer);
}

- (void)testPanGestureRecognizerHasAssociatedViewAfterViewDidLoad {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    __unused id view = photosViewController.view;
    XCTAssertNotNil(photosViewController.panGestureRecognizer.view);
}

- (void)testSingleTapGestureRecognizerExistsAfterInitialization {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    XCTAssertNotNil(photosViewController.singleTapGestureRecognizer);
}

- (void)testSingleTapGestureRecognizerHasAssociatedViewAfterViewDidLoad {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    __unused id view = photosViewController.view;
    XCTAssertNotNil(photosViewController.singleTapGestureRecognizer.view);
}

- (void)testPageViewControllerExistsAfterInitialization {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    [photosViewController viewDidLoad];
    
    XCTAssertNotNil(photosViewController.pageViewController);
}

- (void)testPageViewControllerDoesNotHaveAssociatedSuperviewBeforeViewLoads {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    XCTAssertNil(photosViewController.pageViewController.view.superview);
}

- (void)testPageViewControllerHasAssociatedSuperviewAfterViewLoads {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    photosViewController.view = photosViewController.view; // Referencing the view loads it.
    XCTAssertNotNil(photosViewController.pageViewController.view.superview);
}

- (void)testCurrentlyDisplayedPhotoIsFirstAfterConvenienceInitialization {
    NSArray *photos = [self newTestPhotos];
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:photos];
    [photosViewController viewDidLoad];

    XCTAssertEqualObjects(photos.firstObject, photosViewController.currentlyDisplayedPhoto);
}

- (void)testCurrentlyDisplayedPhotoIsAccurateAfterSettingInitialPhoto {
    NSArray *photos = [self newTestPhotos];
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:photos initialPhoto:photos.lastObject];
    [photosViewController viewDidLoad];

    XCTAssertEqualObjects(photos.lastObject, photosViewController.currentlyDisplayedPhoto);
}

- (void)testCurrentlyDisplayedPhotoIsAccurateAfterDisplayPhotoCall {
    NSArray *photos = [self newTestPhotos];
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:photos initialPhoto:photos.lastObject];
    [photosViewController viewDidLoad];
    [photosViewController displayPhoto:photos.firstObject animated:NO];

    XCTAssertEqualObjects(photos.firstObject, photosViewController.currentlyDisplayedPhoto);
}

- (void)testLeftBarButtonItemIsPopulatedAfterInitialization {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    XCTAssertNotNil(photosViewController.leftBarButtonItem);
}

- (void)testLeftBarButtonItemIsNilAfterSettingToNil {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    photosViewController.leftBarButtonItem = nil;
    XCTAssertNil(photosViewController.leftBarButtonItem);
}

- (void)testLeftBarButtonItemsArePopulatedAfterInitialization {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    XCTAssertNotNil(photosViewController.leftBarButtonItems);
}

- (void)testLeftBarButtonItemsAreNilAfterSettingToNil {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    photosViewController.leftBarButtonItems = nil;
    XCTAssertNil(photosViewController.leftBarButtonItems);
}

- (void)testRightBarButtonItemIsPopulatedAfterInitialization {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    XCTAssertNotNil(photosViewController.rightBarButtonItem);
}

- (void)testRightBarButtonItemIsNilAfterSettingToNil {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    photosViewController.rightBarButtonItem = nil;
    XCTAssertNil(photosViewController.rightBarButtonItem);
}

- (void)testRightBarButtonItemsArePopulatedAfterInitialization {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    XCTAssertNotNil(photosViewController.rightBarButtonItems);
}

- (void)testRightBarButtonItemsAreNilAfterSettingToNil {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    photosViewController.rightBarButtonItems = nil;
    XCTAssertNil(photosViewController.rightBarButtonItems);
}

- (void)testConvenienceInitializerAcceptsNil {
    XCTAssertNoThrow([[NYTPhotosViewController alloc] initWithPhotos:nil]);
}

- (void)testDesignatedInitializerAcceptsNilForPhotosParameter {
    XCTAssertNoThrow([[NYTPhotosViewController alloc] initWithPhotos:nil initialPhoto:[[NYTExamplePhoto alloc] init]]);
}

- (void)testDesignatedInitializerAcceptsNilForInitialPhotoParameter {
    XCTAssertNoThrow([[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos] initialPhoto:nil]);
}

- (void)testDesignatedInitializerAcceptsNilForBothParameters {
    XCTAssertNoThrow([[NYTPhotosViewController alloc] initWithPhotos:nil initialPhoto:nil]);
}

- (void)testDisplayPhotoAcceptsNil {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    XCTAssertNoThrow([photosViewController displayPhoto:nil animated:NO]);
}

- (void)testDisplayPhotoDoesNothingWhenPassedPhotoOutsideDataSource {
    NSArray *photos = [self newTestPhotos];
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:photos initialPhoto:photos.firstObject];
    [photosViewController viewDidLoad];
    
    NYTExamplePhoto *invalidPhoto = [[NYTExamplePhoto alloc] init];
    
    [photosViewController displayPhoto:invalidPhoto animated:NO];
    XCTAssertEqualObjects(photos.firstObject, photosViewController.currentlyDisplayedPhoto);
}

- (void)testDisplayPhotoMovesToCorrectPhoto {
    NSArray *photos = [self newTestPhotos];
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:photos initialPhoto:photos.firstObject];
    [photosViewController viewDidLoad];

    NYTExamplePhoto *photoToDisplay = photos[2];
    
    [photosViewController displayPhoto:photoToDisplay animated:NO];
    XCTAssertEqualObjects(photoToDisplay, photosViewController.currentlyDisplayedPhoto);
}

- (void)testUpdateImageForPhotoAcceptsNil {
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:[self newTestPhotos]];
    XCTAssertNoThrow([photosViewController updateImageForPhoto:nil]);
}

- (void)testUpdateImageForPhotoDoesNothingWhenPassedPhotoOutsideDataSource {
    NSArray *photos = [self newTestPhotos];
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:photos initialPhoto:photos.firstObject];
    NYTExamplePhoto *invalidPhoto = [[NYTExamplePhoto alloc] init];
    invalidPhoto.image = [[UIImage alloc] init];
    
    [photosViewController updateImageForPhoto:invalidPhoto];
    
    XCTAssertEqualObjects(photosViewController.currentlyDisplayedPhoto.image, [photos.firstObject image]);
}

- (void)testUpdateImageForPhotoUpdatesImage {
    NSArray *photos = [self newTestPhotos];
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:photos initialPhoto:photos.firstObject];

    NYTExamplePhoto *photoToUpdate = photos.firstObject;
    photoToUpdate.image = [UIImage imageNamed:@"NYTimesBuilding" inBundle:[NSBundle bundleForClass:[self class]] compatibleWithTraitCollection:nil];
    
    [photosViewController updateImageForPhoto:photoToUpdate];
    
    XCTAssertEqualObjects(photosViewController.currentlyDisplayedPhoto.image, photoToUpdate.image);
}

- (void)testViewIsntLoadedAfterInit {
    NSArray *photos = [self newTestPhotos];
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:photos];

    XCTAssertFalse(photosViewController.isViewLoaded);
}

- (void)testPageViewIsntLoadedAfterInit {
    NSArray *photos = [self newTestPhotos];
    NYTPhotosViewController *photosViewController = [[NYTPhotosViewController alloc] initWithPhotos:photos];

    XCTAssertFalse(photosViewController.pageViewController.isViewLoaded);
}

#pragma mark - Helpers

- (NSArray *)newTestPhotos {
    NSMutableArray *photos = [NSMutableArray array];
    
    for (int i = 0; i < 5; i++) {
        NYTExamplePhoto *photo = [[NYTExamplePhoto alloc] init];
        [photos addObject:photo];
    }
    
    return photos;
}

@end

File: Example/NYTPhotoViewer-SwiftTests/NYTPhotosViewControllerTests.swift
//
//  NYTPhotosViewControllerTests.swift
//  NYTPhotoViewer
//
//  Created by Mark Keefe on 3/21/15.
//  Copyright (c) 2015 The New York Times. All rights reserved.
//

import UIKit
import XCTest

class NYTPhotosViewControllerTests: XCTestCase {

    let photos: [ExamplePhoto] = PhotosProvider().photos
    var photosViewController: NYTPhotosViewController?
    
    override func setUp() {
        super.setUp()
        photosViewController = NYTPhotosViewController(photos: photos)
    }
    
    override func tearDown() {
        photosViewController = nil
        super.tearDown()
    }

    func testPanGestureRecognizerExistsAfterInitialization() {
        XCTAssertNotNil(photosViewController?.panGestureRecognizer)
    }
    
    func testPanGestureRecognizerHasAssociatedView() {
        XCTAssertNotNil(photosViewController?.panGestureRecognizer.view)
    }
    
    func testSingleTapGestureRecognizerExistsAfterInitialization() {
        XCTAssertNotNil(photosViewController?.singleTapGestureRecognizer)
    }
    
    func testSingleTapGestureRecognizerHasAssociatedView() {
        XCTAssertNotNil(photosViewController?.singleTapGestureRecognizer.view)
    }
    
    func testPageViewControllerExistsAfterInitialization() {
        XCTAssertNotNil(photosViewController?.pageViewController)
    }
    
    func testPageViewControllerDoesNotHaveAssociatedSuperviewBeforeViewLoads() {
        XCTAssertNil(photosViewController?.pageViewController.view.superview)
    }
    
    func testPageViewControllerHasAssociatedSuperviewAfterViewLoads() {
        photosViewController?.view = photosViewController?.view // Referencing the view loads it.
        XCTAssertNotNil(photosViewController?.pageViewController.view.superview)
    }
    
    func testCurrentlyDisplayedPhotoIsFirstAfterConvenienceInitialization() {
        XCTAssertEqual(photos.first!, photosViewController?.currentlyDisplayedPhoto as? ExamplePhoto)
    }
    
    func testCurrentlyDisplayedPhotoIsAccurateAfterSettingInitialPhoto() {
        photosViewController = NYTPhotosViewController(photos: photos, initialPhoto: photos.last)
        XCTAssertEqual(photos.last!, photosViewController?.currentlyDisplayedPhoto as? ExamplePhoto)
    }
    
    func testCurrentlyDisplayedPhotoIsAccurateAfterDisplayPhotoCall() {
        photosViewController = NYTPhotosViewController(photos: photos, initialPhoto: photos.last)
        photosViewController?.displayPhoto(photos.first, animated: false)
        XCTAssertEqual(photos.first!, photosViewController?.currentlyDisplayedPhoto as? ExamplePhoto)
    }
    
    func testLeftBarButtonItemIsPopulatedAfterInitialization() {
        XCTAssertNotNil(photosViewController?.leftBarButtonItem)
    }
    
    func testLeftBarButtonItemIsNilAfterSettingToNil() {
        photosViewController?.leftBarButtonItem = nil
        XCTAssertNil(photosViewController?.leftBarButtonItem)
    }
    
    func testRightBarButtonItemIsPopulatedAfterInitialization() {
        XCTAssertNotNil(photosViewController?.rightBarButtonItem)
    }
    
    func testRightBarButtonItemIsNilAfterSettingToNil() {
        photosViewController?.rightBarButtonItem = nil
        XCTAssertNil(photosViewController?.rightBarButtonItem)
    }

    func testConvenienceInitializerAcceptsNil() {
        XCTAssertNotNil(photosViewController)
    }

    func testDesignatedInitializerAcceptsNilForPhotosParameter() {
        photosViewController = NYTPhotosViewController(photos: nil, initialPhoto: photos.first)
        XCTAssertNotNil(photosViewController)
    }

    func testDesignatedInitializerAcceptsNilForInitialPhotoParameter() {
        photosViewController = NYTPhotosViewController(photos: photos, initialPhoto: nil)

        XCTAssertNotNil(photosViewController)
    }
    
    func testDesignatedInitializerAcceptsNilForBothParameters() {
        photosViewController = NYTPhotosViewController(photos: nil, initialPhoto: nil)

        XCTAssertNotNil(photosViewController)
    }
    
    func testDisplayPhotoAcceptsNil() {
        photosViewController?.displayPhoto(nil, animated: false)
        XCTAssertNotNil(photosViewController)
    }
    
    func testDisplayPhotoDoesNothingWhenPassedPhotoOutsideDataSource() {
        photosViewController = NYTPhotosViewController(photos: photos, initialPhoto: photos.first)
        
        let invalidPhoto = ExamplePhoto(attributedCaptionTitle: NSAttributedString(string: "title"))
        
        photosViewController?.displayPhoto(invalidPhoto as NYTPhoto, animated: false)
        XCTAssertEqual(photos.first!, photosViewController?.currentlyDisplayedPhoto as? ExamplePhoto)
    }
    
    func testDisplayPhotoMovesToCorrectPhoto() {
        photosViewController = NYTPhotosViewController(photos: photos, initialPhoto: photos.first)
        let photoToDisplay = photos[2]
        
        photosViewController?.displayPhoto(photoToDisplay, animated: false)
        XCTAssertEqual(photoToDisplay, photosViewController?.currentlyDisplayedPhoto as? ExamplePhoto)
    }
    
    func testUpdateImageForPhotoAcceptsNil() {
        photosViewController?.updateImageForPhoto(nil)
        XCTAssertNotNil(photosViewController)
    }
    
    func testUpdateImageForPhotoDoesNothingWhenPassedPhotoOutsideDataSource() {
        photosViewController = NYTPhotosViewController(photos: photos, initialPhoto: photos[CustomEverythingPhotoIndex])
        let invalidPhoto = ExamplePhoto(attributedCaptionTitle: NSAttributedString())
        invalidPhoto.image = UIImage()
        
        photosViewController?.updateImageForPhoto(invalidPhoto)
        
        if photos.count > CustomEverythingPhotoIndex {
            if let testImage = photos[CustomEverythingPhotoIndex].image {
                /** Swift 1.2
                 *  if photos.count > PhotosProvider.CustomEverythingPhotoIndex,
                 *  let testImage = photos[PhotosProvider.CustomEverythingPhotoIndex].image
                 */
                if let currentlyDisplayedPhoto = photosViewController?.currentlyDisplayedPhoto {
                    XCTAssertEqual(currentlyDisplayedPhoto.image, testImage)
                }
            }
        }
    }
    
    func testUpdateImageForPhotoUpdatesImage() {
        photosViewController = NYTPhotosViewController(photos: photos, initialPhoto: photos.first)

        if let photoToUpdate = photos.first {
            photoToUpdate.image = UIImage()
            
            photosViewController?.updateImageForPhoto(photoToUpdate)
            
            if let currentlyDisplayedPhoto = photosViewController?.currentlyDisplayedPhoto {
                XCTAssertEqual(currentlyDisplayedPhoto.image, photoToUpdate.image!)
            }
        }
    }
}
